proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -104,45 +106,88 @@ f = delegate int { return 1; };         // syntax error
+## Natural (function) type
+A lambda expression has a natural type if the parameters types are explicit and the return type is either explicit or can be inferred (see [inferred return type](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#inferred-return-type)).
 
-Normally method group resolution searches for extension methods lazily, only iterating through successive namespace scopes until extension methods are found that match the target type. But to determine the common natural type will require searching all namespace scopes.
+A method group has a natural type if all candidate methods in the method group have a common signature. (If the method group may include extension methods, the candidates include the containing type and all extension method scopes.)
 
-The requirement of a common type across all methods in the method group means that adding an overload (including an extension method overload) for a method may be a breaking change if the original method was used as a method group with inferred type.
+The natural type of a lambda expression or method group is a _function_type_.
+A _function_type_ represents a method signature: the parameter types and ref kinds, and return type and ref kind.
+Lambda expressions or method groups with the same signature have the same _function_type_.
 
-The delegate type for the lambda or method group and parameter types `P1, ..., Pn` and return type `R` is:
-- if any parameter or return value is not by value, or there are more than 16 parameters, or any of the parameter types or return are not valid type arguments (say, `(int* p) => { }`), then the delegate is a synthesized `internal` anonymous delegate type with signature that matches the lambda or method group, and with parameter names `arg1, ..., argn` or `arg` if a single parameter;
-- if `R` is `void`, then the delegate type is `System.Action<P1, ..., Pn>`;
-- otherwise the delegate type is `System.Func<P1, ..., Pn, R>`.
+A _function_type_ exists at compile time only: _function_types_ do not appear in source or metadata.
 
-The compiler may allow more signatures to bind to `System.Action<>` and `System.Func<>` types in the future (if `ref struct` types are allowed type arguments for instance).
+_Open issue: Should the function_type be available from the compiler API?_
 
-_Should the compiler bind to a matching `System.Action<>` or `System.Func<>` type regardless of arity and synthesize a delegate type otherwise? If so, should the compiler warn if the expected delegate types are missing?_
+### Conversions
+From a _function_type_ `F` there are implicit _function_type_ conversions:
+- To a _function_type_ `G` if the parameters and return types of `F` are variance-convertible to the parameters and return type of `G`
+- To `System.MulticastDelegate` or base classes or interfaces of `System.MulticastDelegate`
+- To `System.Linq.Expressions.Expression` or `System.Linq.Expressions.LambdaExpression`
```

Comment:
>or `System.Linq.Expressions.LambdaExpression` [](http://example.com/codeflow?start=42&length=45)

`System.Linq.Expressions.LambdaExpression` derives from `System.Linq.Expressions.Expression`. Do we need to list both here?